File: Java/BinaryTree/BinaryTree.java
package BinaryTree;

import java.util.ArrayList;
import java.util.Random;

public class BinaryTree {

    private Tree root;

    public static void main(String[] args) {
		BinaryTree binTree = new BinaryTree();
		Random generator = new Random();
		for (int x = 0; x < 1000000; x++) {
			binTree.addLeaf(new AValue(generator.nextInt()));
		}
	}

    public BinaryTree() {
		root = null;
    }

    public void addLeaf(Comparable val) {
		if (root == null) {
			root = new Tree(val);
			return;
		}

		if (val.compareTo(root.value) == 1){
			if (root.right == null) {
				root.right = new Tree(val);
			} else {
				addLeaf(val, root.right);
			}
			return;
		}

		if (root.left == null) {
			root.left = new Tree(val);
		} else {
			addLeaf(val, root.left);
		}
    }

    public void addLeaf(Comparable val, Tree leaf){
		if (val.compareTo(leaf.value) == 1 ){
			if (leaf.right == null){
				leaf.right = new Tree(val);
			} else {
				addLeaf(val, leaf.right);
			}
			return;
		}

		if (leaf.left == null){
			leaf.left = new Tree(val);
		} else {
			addLeaf(val, leaf.left);
		}
    }

    public ArrayList<Comparable> walk(){
		ArrayList<Comparable> listOf = new ArrayList<>();

		if(root == null) {
			return listOf;
		}

		walk(root.left, listOf);
		walk(root.right, listOf);

		return listOf;
    }

    public ArrayList walk(Tree leaf, ArrayList listOf){
		if (leaf == null) {
			return listOf;
		}

		walk(leaf.left, listOf);
		listOf.add(leaf.value);
		walk(leaf.right, listOf);

		return listOf;
    }
}

File: C++/binarytree/binarytree.cpp
#ifndef BTREEINC
#include <binarytree.h>
#define BTREEINC
#endif

template <class T>
BinaryTree<T>::BinaryTree()
	: root(nullptr), levels(0) {}

template <class T>
BinaryTree<T>::~BinaryTree() {
	delete root;
}

template <class T>
unsigned int BinaryTree<T>::Levels() const {
	return levels;
}

template <class T>
void BinaryTree<T>::Add(T i) {

	levels++;

	if (root == nullptr) {
		root = NewNode<T>(i);
		return;
	}

	if (root->Value < i) {
		return add(i, root->Right);
	}
	add(i, root->Left);
	return;
}

template <class T>
void BinaryTree<T>::add(T i, Tree<T>* &leaf) {
	if (leaf == nullptr) {
		leaf = NewNode<T>(i);
		return;
	}
	if (leaf->Value < i) {
		return add(i, leaf->Right);
	}
	return add(i, leaf->Left);
}

template <class T>
vector<T> BinaryTree<T>::Walk() {

	vector<T> out;
	if (root == nullptr) {
		cout << "here\n";
		return out;
	}

	walk(root->Left, out);
	out.push_back(root->Value);
	walk(root->Right, out);

	return out;
}

template <class T>
vector<T> BinaryTree<T>::walk(Tree<T> *leaf, vector<T>& out) {
	if (leaf == nullptr) {
		return out;
	}

	walk(leaf->Left, out);
	out.push_back(leaf->Value);
	walk(leaf->Right, out);

	return out;
}

